proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+  /// this is an unsafe operation.
+  ///
+  /// - Parameters:
+  ///   - offset: The offset from this pointer, in bytes. `offset` must be
+  ///     nonnegative. The default is zero.
+  ///   - type: The type of the instance to create.
+  /// - Returns: A new instance of type `T`, read from the raw bytes at
+  ///     `offset`. The returned instance isn't associated
+  ///     with the value in the range of memory referenced by this pointer.
+  public func loadUnaligned<T: BitwiseCopyable>(
+    fromUncheckedByteOffset offset: Int, as: T.Type
+  ) -> T
+}
+```
+
+A `RawSpan` can be viewed as a `Span<T>`, provided the memory is laid out homogenously as instances of `T`.
+
+```swift
+  /// View the memory span represented by this view as a different type
+  ///
+  /// The memory must be laid out identically to the in-memory representation of `T`.
+  ///
+  /// - Parameters:
+  ///   - type: The type you wish to view the memory as
+  /// - Returns: A new `Span` over elements of type `T`
+  public func view<T: BitwiseCopyable>(as: T.Type) -> dependsOn(self) Span<T>
+}
+```
+
+##### Index-related Operations:
+
+```swift
+extension RawSpan {
+  /// The number of bytes in the span.
+  public var count: Int { get }
+
+  /// A Boolean value indicating whether the span is empty.
+  public var isEmpty: Bool { get }
+
+  /// The indices that are valid for subscripting the span, in ascending
+  /// order.
+  public var indices: Range<Int> { get }
+
+  /// Traps if `offset` is not a valid offset into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - position: an offset to validate
+  public func boundsCheckPrecondition(_ offset: Int)
+
+  /// Traps if `bounds` is not a valid range of offsets into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offsets: a range of offsets to validate
+  public func boundsCheckPrecondition(_ offsets: Range<Int>)
+}
+```
+
+##### Index validation utiliities:
+
+Every time `RawSpan` uses an index or an integer offset, it checks for their validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension RawSpan {
+  /// Traps if `position` is not a valid index for this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - position: an Index to validate
+  public boundsCheckPrecondition(_ position: Index)
+
+  /// Traps if `bounds` is not a valid range of indices for this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - bounds: a range of indices to validate
+  public boundsCheckPrecondition(_ bounds: Range<Index>)
+
+  /// Traps if `offset` is not a valid offset into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offset: an offset to validate
+  public boundsCheckPrecondition(offset: Int)
+
+  /// Traps if `offsets` is not a valid range of offsets into this `RawSpan`
+  ///
+  /// - Parameters:
+  ///   - offsets: a range of offsets to validate
+  public boundsCheckPrecondition(offsets: Range<Int>)
+}
+```
```

Comment:
This section is redundant with the functions in the prior one

Reply:
I'll remove this.